src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestStream.cs

```diff
@@ -156,30 +224,41 @@ private async ValueTask<int> ReadAsyncInternal(Memory<byte> buffer, Cancellation
 
                 var consumed = readableBuffer.End;
                 var actual = 0;
+
+                var advanced = false;
                 try
                 {
                     if (readableBufferLength != 0)
                     {
-                        actual = (int)Math.Min(readableBufferLength, buffer.Length);
+                        actual = (int)Math.Min(readableBufferLength, reader.Buffer.Length);
 
                         var slice = actual == readableBufferLength ? readableBuffer : readableBuffer.Slice(0, actual);
                         consumed = slice.End;
-                        slice.CopyTo(buffer.Span);
+                        slice.CopyTo(reader.Buffer.Span);
 
-                        return actual;
+                        // Finally blocks in enumerators aren't excuted prior to the yield return,
+                        // so we advance here
```

Comment:
Hm... that seems like it will be a bug farm in the future.

Reply:
I don't think this is that bad. The finally still runs given an uncaught exception or the completion of the method. I think it would be more weird if the finally did run for every yield return.